Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,363 @@
+PR #12107). With this in mind we will sometimes refer to objects and morphisms in monoidal
+categories as 1-morphisms and 2-morphisms respectively, borrowing the terminology of bicategories.
+Note that the relation between monoidal categories and bicategories is formalized in
+`Mathlib.CategoryTheory.Bicategory.SingleObj`, although the string diagram widget does not use
+it directly.
+
+-/
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab
+open CategoryTheory
+open Mathlib.Tactic.Coherence
+
+open Mathlib.Tactic.Monoidal
+
+namespace Widget.StringDiagram
+
+/-! ## Objects in string diagrams -/
+
+/-- Nodes for 2-morphisms in a string diagram. -/
+structure AtomNode : Type where
+  /-- The vertical position of the node in the string diagram. -/
+  vPos : ℕ
+  /-- The horizontal position of the node in the string diagram, counting strings in domains. -/
+  hPosSrc : ℕ
+  /-- The horizontal position of the node in the string diagram, counting strings in codomains. -/
+  hPosTar : ℕ
+  /-- The underlying expression of the node. -/
+  atom : Atom
+
+/-- Nodes for identity 2-morphisms in a string diagram. -/
+structure IdNode : Type where
+  /-- The vertical position of the node in the string diagram. -/
+  vPos : ℕ
+  /-- The horizontal position of the node in the string diagram, counting strings in domains. -/
+  hPosSrc : ℕ
+  /-- The horizontal position of the node in the string diagram, counting strings in codomains. -/
+  hPosTar : ℕ
+  /-- The underlying expression of the node. -/
+  id : Atom₁
+
+/-- Nodes in a string diagram. -/
+inductive Node : Type
+  | atom : AtomNode → Node
+  | id : IdNode → Node
+
+/-- The underlying expression of a node. -/
+def Node.e : Node → Expr
+  | Node.atom n => n.atom.e
+  | Node.id n => n.id.e
+
+/-- The domain of the 2-morphism associated with a node as a list
+(the first component is the node itself). -/
+def Node.srcList : Node → MetaM (List (Node × Atom₁))
+  | Node.atom n => return (← n.atom.src).toList.map (fun f ↦ (.atom n, f))
+  | Node.id n => return [(.id n, n.id)]
+
+/-- The codomain of the 2-morphism associated with a node as a list
+(the first component is the node itself). -/
+def Node.tarList : Node → MetaM (List (Node × Atom₁))
+  | Node.atom n => return (← n.atom.tgt).toList.map (fun f ↦ (.atom n, f))
+  | Node.id n => return [(.id n, n.id)]
+
+/-- The vertical position of a node in a string diagram. -/
+def Node.vPos : Node → ℕ
+  | Node.atom n => n.vPos
+  | Node.id n => n.vPos
+
+/-- The horizontal position of a node in a string diagram, counting strings in domains. -/
+def Node.hPosSrc : Node → ℕ
+  | Node.atom n => n.hPosSrc
+  | Node.id n => n.hPosSrc
+
+/-- The horizontal position of a node in a string diagram, counting strings in codomains. -/
+def Node.hPosTar : Node → ℕ
+  | Node.atom n => n.hPosTar
+  | Node.id n => n.hPosTar
+
+/-- The list of nodes at the top of a string diagram. -/
+def topNodes (η : WhiskerLeftExpr) : MetaM (List Node) := do
```

Comment:
Similarly, here and below, there is probably a performance benefit swapping out `List` for `Array` but I don't know what size the diagrams need to be before it is realized.

Reply:
I would say no diagram you'd want to see in practice is large enough for this to make a meaningful difference. But I haven't benchmarked it, only tasseomancied.

Reply:
I agree. I think the direct `Expr` manipulation might be the best place to look for a real world improvement. This is also something that can be thought about afterwards.